### PR TITLE
Implement sound is_affine() test

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #72 

Recall that our affine pattern matching algorithm consists of three steps:
1. (#245) an optimistic (necessarily incomplete) unsound pattern matching step to match possibly affine funsors; 
2. (this PR) a sound but incomplete `is_affine()` test to eliminate non-affine funsors;
3. (#249) an algorithm to substitute in probe funsors and extract affine coefficients from a funsor known to be affine.

This also removes the obsolete `.is_affine` attribute of `Contraction`.

## Tested

- positive cases are exercised by existing tests
- added new positive check and cases to `test_extract_affine()`
- added a new unit test for negative cases

### CODE REVIEW

**Comment by fritzo** [comment]
cc @fehiepsi in case you are interested.

File: funsor/affine.py [hunk 1]
```diff
@@ -1,16 +1,103 @@
 from collections import OrderedDict
+from functools import reduce
 
 import opt_einsum
 import torch
+from multipledispatch import dispatch
```
Comment (eb8680) [inline funsor/affine.py]: nit: use `singledispatch` here instead?  I think it's faster

File: test/test_affine.py [hunk 2]
```diff
@@ -116,3 +119,21 @@ def test_extract_affine(expr):
                          for k, (coeff, eqn) in coeffs.items())
     assert isinstance(actual, Tensor)
     assert_close(actual, expected)
+
+
+@pytest.mark.parametrize("expr", [
```
Comment (eb8680) [inline test/test_affine.py]: Add some test cases covering `Unary`?

**Comment by fritzo** [comment]
> Can you add some positive test cases ... with random substitutions ...

Yes, I forgot to mention positive tests are tested by `test_extract_affine()` via random substitution.

File: test/test_affine.py (hunk 2 above)
Comment (fritzo) [reply test/test_affine.py]: done.

File: funsor/affine.py (hunk 1 above)
Comment (fritzo) [reply funsor/affine.py]: done.

**Comment by fritzo** [comment]
Thanks for the careful review!

**Pull request merged**